src/goto-instrument/code_contracts.cpp

```diff
@@ -282,24 +335,388 @@ const symbolt &code_contractst::new_tmp_symbol(
+          //   be able to assign, we need to create ghost variables to store
+          //   references to that memory.
+          // If a malloc is in a loop, we can't generally determine how many
+          //   times it will run in order to create an appropriate number of
+          //   references for the assignable memory.
+          // So, if we have a malloc in a loop, we can't statically create the
+          //   assertion statements needed to enforce the assigns clause.
+          log.error() << "Call to malloc at location "
+                      << malloc_entry.location_number << " falls between goto "
+                      << "source location " << goto_entry.location_number
+                      << " and it's destination at location "
+                      << target->location_number << ". "
+                      << "Unable to complete instrumentation, as this malloc "
+                      << "may be in a loop." << messaget::eom;
+          throw 0;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+bool code_contractst::add_pointer_checks(const std::string &function_name)
+{
+  // Get the function object before instrumentation.
+  auto old_function = goto_functions.function_map.find(function_name);
+  if(old_function == goto_functions.function_map.end())
+  {
+    log.error() << "Could not find function '" << function_name
+                << "' in goto-program; not enforcing contracts."
+                << messaget::eom;
+    return true;
+  }
+  goto_programt &program = old_function->second.body;
+  if(program.instructions.empty()) // empty function body
+  {
+    return false;
+  }
+
+  const irep_idt function_id(function_name);
+  const symbolt &function_symbol = ns.lookup(function_id);
+  const code_typet &type = to_code_type(function_symbol.type);
+
+  exprt assigns = static_cast<const exprt &>(type.find(ID_C_spec_assigns));
+
+  // Return if there are no reference checks to perform.
+  if(assigns.is_nil())
+    return false;
+
+  goto_programt::instructionst::iterator instruction_iterator =
+    program.instructions.begin();
+
+  // Create temporary variables to hold the assigns clause targets before
+  // they can be modified.
+  goto_programt standin_decls;
+  std::vector<exprt> original_references;
+  populate_assigns_references(
+    function_symbol, function_id, standin_decls, original_references);
+
+  // Create a list of variables that are okay to assign.
+  std::set<exprt> freely_assignable_exprs;
+  for(code_typet::parametert param : type.parameters())
+  {
+    freely_assignable_exprs.insert(param);
+  }
+
+  int lines_to_iterate = standin_decls.instructions.size();
+  program.insert_before_swap(instruction_iterator, standin_decls);
+  std::advance(instruction_iterator, lines_to_iterate);
+
+  if(check_for_looped_mallocs(program))
+  {
+    return true;
+  }
+
+  // Insert aliasing assertions
+  for(; instruction_iterator != program.instructions.end();
+      std::advance(instruction_iterator, 1))
```

Comment:
Why would `++instruction_iterator` not be sufficient here?

Reply:
Implemented as suggested.